src/model.rs

```diff
@@ -606,6 +606,192 @@ impl Collections {
+        // This function finds the largest and returns the two bounds/index (23 and 3 in the first example)
+        fn find_main_hole_boundaries(timetable: &TimeTable, hours_per_day: u8) -> Option<(u8, u8)> {
+            let mut holes: Vec<Vec<u8>> = Vec::new();
+            let mut is_last_elem_hole = false;
+            for i in 0..hours_per_day {
+                if !&timetable.contains_key(&i) {
+                    if !is_last_elem_hole {
+                        holes.push(vec![i]);
+                        is_last_elem_hole = true;
+                    } else if let Some(last) = holes.last_mut() {
+                        last.push(i)
+                    }
+                    // for the midnight passing, concatenate the vectors
+                    // [0,1] and [22,23] will become [22,23,0,1]
+                    if i == hours_per_day - 1 && holes.len() > 1 && holes[0].contains(&0) {
+                        let hole0 = holes[0].clone();
+                        if let Some(last) = holes.last_mut() {
+                            last.extend_from_slice(&hole0)
+                        }
+                        holes.remove(0);
+                    }
+                } else {
+                    is_last_elem_hole = false;
+                }
+            }
+            // *** first, sorts in descending order of width
+            // [[9,10],[12,13,14],[23,0,1]] --> [[12,13,14],[23,0,1],[9,10]]
+            // *** then, in case of equal width, sorts by taking the segment early in the morning (smallest index)
+            // --> [[23,0,1],[12,13,14],[9,10]]
+            holes.sort_unstable_by(|l, r| l.iter().min().cmp(&r.iter().min()));
+            holes.sort_unstable_by(|l, r| r.len().cmp(&l.len()));
+            holes.first().and_then(|mh| {
+                let first_idx = mh.first();
+                let last_idx = mh.last();
+                match (first_idx, last_idx) {
+                    (Some(first_idx), Some(last_idx)) => {
+                        Some((first_idx.to_owned(), last_idx.to_owned()))
+                    }
+                    _ => None,
+                }
+            })
+        }
+
+        // Check if there is a need to calculate the opening/closing times.
+        // Generally it should be all or nothing (absent from a Gtfs, normally present if it's a recent Ntfs)
+        // In all cases a 2nd check is made below
+        let mut required_operation = false;
+        for line in self.lines.values() {
+            if line.opening_time.is_none() || line.closing_time.is_none() {
+                required_operation = true;
+                break;
+            }
+        }
+        if required_operation {
+            let vjs_by_line = get_vjs_by_line(&self);
+            let mut lines = self.lines.take();
+            for line in &mut lines {
+                // 2nd check (see above) to avoid overwriting line opening/closing
+                if line.opening_time.is_none() || line.closing_time.is_none() {
```

Comment:
You're doing twice the same test so maybe a closure can be extracted?

```rust
let check_time_empty = |line| line.opening_time.is_none() || line.closing_time.is_none();
```

Reply:
Done in 2afc55e3d3f76d9b27d27691f8814870618710f7